great_expectations/expectations/metrics/util.py

```diff
@@ -308,13 +322,62 @@ def attempt_allowing_relative_error(dialect):
     return detected_redshift or detected_psycopg2
 
 
+class CaseInsensitiveString(str):
+    """
+    A string that compares equal to another string regardless of case,
+    unless it is quoted.
+    """
+
+    def __init__(self, string: str):
+        # TODO: check if string is already a CaseInsensitiveString?
+        self._original = string
+        self._lower = string.lower()
+        self._quote_string = '"'
+
+    @override
+    def __eq__(self, other: CaseInsensitiveString | str | object):
+        if self.is_quoted():
+            return self._original == str(other)
+        if isinstance(other, CaseInsensitiveString):
+            return self._lower == other._lower
+        elif isinstance(other, str):
+            return self._lower == other.lower()
+        else:
+            return False
+
+    def __hash__(self):
+        return hash(self._lower)
+
+    @override
+    def __str__(self) -> str:
+        return self._original
+
+    def is_quoted(self):
+        return self._original.startswith(self._quote_string)
+
+
+class CaseInsensitiveNameDict(UserDict):
+    """Normal dict except it returns a case-insensitive string for any `name` key values."""
+
+    def __init__(self, data: dict[str, Any]):
+        self.data = data
+
+    @override
+    def __getitem__(self, key: Any) -> Any:
+        item = self.data[key]
+        if key == "name":
+            logger.debug(f"CaseInsensitiveNameDict.__getitem__ - {key}:{item}")
+            return CaseInsensitiveString(item)
+        return item
+
+
 def get_sqlalchemy_column_metadata(
     execution_engine: SqlAlchemyExecutionEngine,
     table_selectable: sqlalchemy.Select,
     schema_name: Optional[str] = None,
-) -> Optional[List[Dict[str, Any]]]:
+) -> Sequence[Mapping[str, Any]] | None:
```

Comment:
The fact that I could change this to an immutable type means that we must not be directly mutating this return value.
Either that or our type-checking coverage is still too bad 😄